Fix #269467: added support to ignore diacritic accents

finleylau opened the pull request:
Used regex to replace special characters with accents, acutes, grave, umlaut, etc. with corresponding English characters

Jojo-Schmitz commented on mscore/instrwidget.cpp:
```diff
@@ -56,6 +56,26 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
+                  text = text.replace(QRegularExpression("[\\x{00D1}]"), QChar('n'));
+                  text = text.replace(QRegularExpression("[\\x{00DD}]"), QChar('y'));
+
                   bool isMatch = text.contains(searchPhrase, Qt::CaseInsensitive);
```
Instead you could directly use a `contains()` variant that works with regular expressions, like `bool QString::contains(const QRegularExpression &re) const`.
Also what about e.g. the Dutch `ij` and the German `ß`. Which characters are you covering, can you add a comment about that, e.g. mentioning their HTML entity, like `&auml;` for `ä`?

lasconic commented:
I like this approach better https://stackoverflow.com/questions/12278448/removing-accents-from-a-qstring#12806471 but I don't know how it would behave for CJK, russian, arabic, hebrew and others.

finleylau commented:
For non-Latin languages, is the goal to allow them to type Latin characters and have for example Russian characters appear?  or is the goal to find a way to replace accented Latin characters with unaccented ones, but still leave the non-Latin characters the same?

lasconic commented:
I would leave non Latin languages alone, I assume someone using MuseScore in russian can type russian and doesn't need transliteration.

finleylau commented:
I've implemented it currently to the first solution suggested in https://stackoverflow.com/questions/12278448/removing-accents-from-a-qstring#12806471.  The answer posted replaces too many characters and potentially could mess up many non Latin languages.  The code is now decomposing the Unicode, making sure that each Unicode character is not an accent, and then recomposing the Unicode value.  However, it doesn't work for the ß or æ or similar characters, as  these characters are considered by the QT Unicode system as being "Lower Case Letters" and therefore can't programatically be decomposed and recomposed like other accented characters.

One solution would be to leave them in original form and allow users to just type the nonLatin characters and leave the solution as it stands now.  Another more complicated solution would be to manually replace characters like Ø and the above mentioned characters and allow for both typing the nonLatin characters OR typing the corresponding Latin characters.  However, I'm not aware of the correct Latin spelling, for example I've read that ß can be represented by either 'ss', 'sz', or just 's'.  Do you have any suggestions?

Jojo-Schmitz commented on mscore/instrwidget.cpp:
```diff
@@ -56,7 +56,45 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
+                                    }
+                              else if (text.at(i) == QChar(0x00E6)) {
+                                    // ae, &#230
+                                    manualReplaceDiacritic.append(QString("ae"));
```
There's also äüöÄÜÖ, transcribed as ae, oe. ue, Ae, Oe, Ue, and also there's œ, Œ, and Æ, transcribed as oe, OE, AE
There may be more

Jojo-Schmitz commented on mscore/instrwidget.cpp:
```diff
@@ -56,7 +56,45 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
+                                    }
+                              else if (text.at(i) == QChar(0x00DF)) {
+                                    // eszett, &#223
+                                    manualReplaceDiacritic.append(QString("ss"));
```
indeed the correct transcription, even if it gets pronounced as "sz" or "sharp s"

lasconic commented:
*should i do them in mtest/scripting/tst_scripting.cpp  or should i make a tst stringutils project in libmscore*

I would create a new mtest/stringutils/tst_stringutils.cpp file

finleylau commented:
Added the stringutils class and added three tests (normal string that shouldn't be modified, one to replace all the ligatures, and one to replace the diacritics)

Jojo-Schmitz commented:
That merge commit shouldn't be there, don't merge, but rebase

finleylau commented:
Continued in #3496
